File: src/Receiver/connection.py
from pynput import mouse as ms, keyboard as kb
import screeninfo as si
import socket
import platform
import random
import time
import getmac

BUFFER = 1024
DISCOVER = 'DISCOVER_SERVER'
AVAILABLE = 'SERVER_AVAILABLE'
REQUEST = 'REQUEST_CONNECTION'
ACCEPT = 'CONNECTION_ACCEPTED'
DENY = 'CONNECTION_DENIED'

screenWidth, screenHeigth = si.get_monitors()[0].width, si.get_monitors()[0].height

udp = None
mouse = ms.Controller()


class Connection():
    hostname = socket.gethostname()
    local_ip = socket.gethostbyname(hostname)
    system = platform.system().lower()
    mac = getmac.get_mac_address().upper()

    addr = "0.0.0.0"
    port = 5000
    message = []
    response = ""
    passcode = ""

    connected = False
    trustedDevice = False
    canCommand = False

    @staticmethod
    def connectToReceiver():
        global udp, canCommand

        udp = socket.socket(
            family=socket.AF_INET,
            type=socket.SOCK_DGRAM
        )
        udp.setsockopt(
            socket.SOL_SOCKET,
            socket.SO_BROADCAST,
            1
        )
        udp.bind((
            "0.0.0.0",
            Connection.port
        ))

        print (f"Waiting for receivers...\n")
        while True:
            try:
                data, Connection.addr = udp.recvfrom(BUFFER)
                message = data.decode('utf-8').split("|")

                if message[0] == DISCOVER:
                    print (f"{DISCOVER} received from {Connection.addr[0]}")
                    response = f"{AVAILABLE}|{Connection.hostname}|{Connection.system}"
                    udp.sendto(
                        response.encode('utf-8'),
                        Connection.addr
                    )
                    print (f"{AVAILABLE} sent to {Connection.addr[0]}\n")
                    Connection.passcode = str(random.randint(1000, 9999))
                    print (f"Use passcode {Connection.passcode} in your Android to connect\n")

                elif message[0] == REQUEST:
                    print (f"{REQUEST} received from {Connection.addr[0]}")

                    if message[1] == Connection.passcode:
                        Connection.port = random.randint(5001, 8000)
                        currentMouseX, currentMouseY = mouse.position
                        response = f"{ACCEPT}|{Connection.mac}|{Connection.port}|{screenWidth}|{screenHeigth}|{currentMouseX}|{currentMouseY}"
                        udp.sendto(
                            response.encode('utf-8'),
                            Connection.addr
                        )

                        print (f"{ACCEPT} sent to {Connection.addr[0]}\n")
                        udp.close()
                        udp = socket.socket(
                            socket.AF_INET,
                            socket.SOCK_DGRAM
                        )
                        udp.bind(("0.0.0.0", Connection.port))
                        print (f"Connected to {Connection.addr[0]}\n\n")
                        Connection.connected = True
                        Connection.canCommand = True
                        break
                    else:
                        print (f"INVALID PASSCODE RECEIVED FROM {Connection.addr[0]}\n")
                        print (f"Use passcode {Connection.passcode} in your Android to connect\n\n")

            except socket.error as e:
                print (e)

    @staticmethod
    def receiveCommands():
        global udp

        while True:
            if(Connection.canCommand):
                data, Connection.addr = udp.recvfrom(BUFFER)
                message = data.decode('utf-8').split("|")
                
                match(message[2]):
                    case "move":
                        mouse.position = (float(message[0]), float(message[1]))
                       
                    case "click":
                        mouse.click(ms.Button.left, 1)
                    
                    case "double":
                        mouse.click(ms.Button.left, 2)
                    
                    case "right":
                        mouse.click(ms.Button.right, 1)
                

                time.sleep(0.01)

File: src/Receiver/lib/connection.dart
import 'importManager.dart';

class Connection {
  static late RawDatagramSocket udp;
  static late Datagram datagram;
  static late Point<int>? startMousePosition;

  static bool trustedDevice = false;
  static bool connected = false;

  static String addr = "";
  static List<String> message = [];   
  static String response = "";
  static String status = "Waiting for connection..."; 

  static int port = 5000;
  static final int BUFFER = 1024;

  static final String DISCOVER = 'DISCOVER_SERVER';
  static final String AVAILABLE = 'SERVER_AVAILABLE';
  static final String REQUEST = 'REQUEST_CONNECTION';
  static final String ACCEPT = 'CONNECTION_ACCEPTED';
  static final String DENY = 'CONNECTION_DENIED';

  static void generatePasscode() {
    const numbers = '0123456789';
    final random = Random();

    String randomChar(String chars) => chars[random.nextInt(chars.length)];

    String newPasscode = 
        '${randomChar(numbers)}${randomChar(numbers)}${randomChar(numbers)}${randomChar(numbers)}';

    Home.passcode = newPasscode;
  }

  static Future<void> connectToReceiver() async {
    udp = await RawDatagramSocket.bind(InternetAddress.anyIPv4, port);
    udp.broadcastEnabled = true;

    await for (RawSocketEvent event in udp) {
      if (event == RawSocketEvent.read) {
        datagram = udp.receive()!;
        message = utf8.decode(datagram.data).split("|");

        if (message[0] == DISCOVER) {
          response = "$AVAILABLE|${Home.hostname}|${Home.system}";

          udp.send(
            utf8.encode(response),
            InternetAddress("255.255.255.255"),
            port,
          );

        } else if (message[0] == REQUEST) {
          if (message[1] == Home.passcode) {
            addr = datagram.address.address;
            port = 5001 + Random().nextInt(2999);
            startMousePosition = await FlutterAutoGUI.position();

            response = "$ACCEPT|$port|${startMousePosition?.x}|${startMousePosition?.y}";

            udp.send(
              utf8.encode(response),
              InternetAddress(addr),
              5000,
            );

            udp.close();
            Connection.connected = true;
            await commandStream();
          }
        }
      }
    }
  }

  static Future<void> commandStream() async {
    udp = await RawDatagramSocket.bind(
      InternetAddress.anyIPv4,
      port,
    );

    while (Connection.connected) {
      await for (RawSocketEvent event in udp) {
        if (event == RawSocketEvent.read) {
          datagram = udp.receive()!;
          message = utf8.decode(datagram.data).split("|");

          switch (message[2]) {
            case 'move':
              FlutterAutoGUI.moveTo(
                point: Point(
                  int.parse(message[0]),
                  int.parse(message[1]),
                ),
              );
              break;

            case 'click':
              FlutterAutoGUI.click(
                button: MouseButton.left,
                clicks: 1,
              );
              break;

            case 'double':
              FlutterAutoGUI.click(
                button: MouseButton.left,
                clicks: 2,
              );
              break;

            case 'right':
              FlutterAutoGUI.click(
                button: MouseButton.right,
                clicks: 1,
              );
              break;
          }
          await Future.delayed(Duration(milliseconds: 10));
        }
      }
    }
  }

  static void closeConnection() {
    udp.close();
    Connection.connected = false;
  }
}
